### PR TITLE
Do exponential backoff in clientset sync

### PR DESCRIPTION
**Pull request by caesarxuchao**
... to avoid hot loops.

First commit is #52.

### CODE REVIEW

File: pkg/agent/agentclient/clientset.go [hunk 1]
```diff
@@ -124,27 +124,37 @@ func (cs *ClientSet) newAgentClient() (*AgentClient, error) {
...
 		if cs.serverCount != 0 {
-			sleep := cs.syncInterval + time.Duration(rand.Float64()*jitter*float64(cs.syncInterval))
-			time.Sleep(sleep)
+			time.Sleep(backoff.Duration)
 		}
 		if cs.serverCount == 0 || cs.ClientsCount() < cs.serverCount {
 			c, err := cs.newAgentClient()
 			if err != nil {
 				klog.Error(err)
+				backoff.Step()
```
[the remark this answers is not in the document]
Comment (caesarxuchao) [reply pkg/agent/agentclient/clientset.go]: Good catch. I accidentally deleted the "Factor" line. Added back.

**Comment by caesarxuchao** [comment]
>  Should this scenario warrant an exponential backoff?

No. We should only exponential backoff when we fail to connect to any server. Updated. Thanks.

File: pkg/agent/agentclient/clientset.go [hunk 2]
```diff
@@ -124,20 +124,30 @@ func (cs *ClientSet) newAgentClient() (*AgentClient, error) {
 	return newAgentClient(cs.address, cs.agentID, cs, cs.dialOption)
 }
 
+func (cs *ClientSet) resetBackoff() *wait.Backoff {
+	return &wait.Backoff{
+		Steps:    3,
+		Jitter:   0.1,
```
Comment (Jefftree) [inline pkg/agent/agentclient/clientset.go]: It still looks like this parameter is not used, should we remove it?

**Comment by caesarxuchao** [comment]
@Jefftree I refactored the code. PTAL. Thank you.

File: pkg/agent/agentclient/clientset.go [hunk 3]
```diff
@@ -128,30 +128,48 @@ func (cs *ClientSet) newAgentClient() (*AgentClient, error) {
...
+		Factor:   1.5,
+		Duration: cs.syncInterval,
+	}
```
Comment (dberkov) [inline pkg/agent/agentclient/clientset.go]: Do you want initialize "Cap" too, just in  case it grows too much?

File: pkg/agent/agentclient/clientset.go [hunk 4]
```diff
@@ -128,30 +128,48 @@ func (cs *ClientSet) newAgentClient() (*AgentClient, error) {
...
-				continue
-			}
-			klog.Infof("sync added client connecting to proxy server %s", c.stream.serverID)
-			go c.Serve()
+		if err := cs.syncOnce(); err != nil {
+			klog.Error(err)
+			duration = backoff.Step()
+		} else {
+			backoff = cs.resetBackoff()
+			duration = wait.Jitter(backoff.Duration, backoff.Jitter)
```
Comment (dberkov) [inline pkg/agent/agentclient/clientset.go]: based on Jitter documentation, we can set second parameter as 0.0, so default value will be selected

File: pkg/agent/agentclient/clientset.go (hunk 3 above)
Comment (Jefftree) [reply pkg/agent/agentclient/clientset.go]: +1

Comment (caesarxuchao) [reply pkg/agent/agentclient/clientset.go]: Done

File: pkg/agent/agentclient/clientset.go (hunk 4 above)
Comment (caesarxuchao) [reply pkg/agent/agentclient/clientset.go]: I think using the value defined in this file is more readable than using the default value.

**Comment by caesarxuchao** [comment]
PTAL. Thank you for the reviews!

**Pull request merged**